[SCEV] forgetValue: support (with-overflow-inst op0, op1)

v01dXYZ opened the pull request:
Fixes #97586.

~The fix is not satisfactory as there is code duplication and it could get out of sync with `MatchBinaryOp` in the future if new patterns are added. The useful part of this commit is the test.~

Commit:
(Draft) [SCEV] forgetValue: support (extractvalue 0, (with-overflow-inst op0, op1))

Without that, forgetValue stops at (with-overflow-inst op0, op1) while
thanks to MatchBinaryOp, SCEV creation considers %extractvalue as if
it is (op op0, op1).

Because of that, it creates an unclearable SCEV value that could
possibly turn out of sync after a transform is applied.

The commit is in Draft as the fix is not satisfactory (code
duplication, do we push EVO to Visited ?). Meant for discussion and
for adding a test.

Nirhar commented:
@v01dXYZ Did you verify if your patch fixes the issue #97586 ?

v01dXYZ commented:
Yes, I verified with the input you provided in your issue. The program doesn't abort.

<details>
<summary>
Input
</summary>

``` llvm
; ModuleID = '<stdin>'
source_filename = "<stdin>"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128-ni:1-p2:32:8:8:32-ni:2"
target triple = "x86_64-unknown-linux-gnu"

define void @ham() {
bb:
  br label %bb1

bb1:                                              ; preds = %bb3, %bb
  %phi = phi i32 [ 0, %bb ], [ %phi4, %bb3 ]
  br label %bb5

bb2:                                              ; preds = %bb7
  %call = call { i32, i1 } @llvm.smul.with.overflow.i32(i32 %add8, i32 -2)
  %extractvalue = extractvalue { i32, i1 } %call, 0
  br label %bb3

bb3:                                              ; preds = %bb3, %bb2
  %phi4 = phi i32 [ %add, %bb3 ], [ 0, %bb2 ]
  %add = add i32 %extractvalue, %phi4
  br i1 false, label %bb3, label %bb1

bb5:                                              ; preds = %bb7, %bb1
  %phi6 = phi i32 [ 1, %bb1 ], [ 0, %bb7 ]
  %icmp = icmp eq i32 %phi, 0
  br i1 %icmp, label %bb9, label %bb7

bb7:                                              ; preds = %bb5
  %load = load i32, ptr addrspace(1) null, align 4
  %add8 = add i32 %load, 1
  br i1 false, label %bb2, label %bb5

bb9:                                              ; preds = %bb5
  ret void
}

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare { i32, i1 } @llvm.smul.with.overflow.i32(i32, i32) #0

attributes #0 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
```

``` sh
~/llvm-project/build.indvars_simplify/bin/opt <filename.ll> \
    -S -debug --print-before-all \
    --passes="loop-unroll<no-partial;peeling;no-runtime;no-upperbound;no-profile-peeling;full-unroll-max=0;O3>" 
```
</details>

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -8412,8 +8412,35 @@ void ScalarEvolution::visitAndClearUsers(
   while (!Worklist.empty()) {
     Instruction *I = Worklist.pop_back_val();
-    if (!isSCEVable(I->getType()))
+    if (!isSCEVable(I->getType())) {
```
```suggestion
    if (!isSCEVable(I->getType()) && !isa<WithOverflowInst>(I)) {
```
should be sufficient, no?

nikic commented on llvm/unittests/Analysis/ScalarEvolutionTest.cpp:
```diff
@@ -1589,4 +1589,40 @@ TEST_F(ScalarEvolutionsTest, ApplyLoopGuards) {
+      "exit: "
+      "  ret void "
+      "} ",
+      Err, C);
```
lit test please.

v01dXYZ replied:
If I only add a lit test, how can I be certain it will exactly test the eviction and not the fact that SCEV value was never created or the stale value used because the implementation of a pass has changed ?

I can add the case from the issue but it requires `LoopPeel` to produce a CFG that is out of sync with the SCEV value.

I don't know how to have a robust test for something as specific using an integration test.

v01dXYZ replied:
thanks. I'll do that.

Commit:
[LoopUnroll] pre-commit test for LoopPeel SCEV invalidation

Warning this test passes only with asserts disabled.
Otherwise the following assertion failure occurs:

Assertion `isAvailableAtLoopEntry(Operands[i], L) && "SCEVAddRecExpr operand is not available at loop entry!"' failed.

Commit:
[LoopUnroll] test for LoopPeel SCEV invalidation

Commit:
[SCEV] visitAndClearUsers support WithOverflowInst more simply

nikic commented on llvm/test/Transforms/LoopUnroll/peel-loop-scev-invalidate-with-overflow-inst.ll:
```diff
@@ -0,0 +1,76 @@
+; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
+declare { i32, i1 } @llvm.smul.with.overflow.i32(i32, i32) #0
+
+attributes #0 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
```
Can drop the intrinsic declaration and attributes.

nikic reviewed (approved):
LGTM
